### PR TITLE
Adding initial version of howto-menu

### PR DESCRIPTION
**Pull request by devnook**

### CODE REVIEW

File: elements/howto-menu/demo.html [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  howto-menuitem {
+    display: inline-block;
+    width: 100%;
+    padding: .2em .5em;
+    border: 1px solid lightgrey;
+  }
+
+  /* The menu is closed. */
+  howto-menu[aria-hidden="true"] {
+    visibility: hidden;
```
Comment (surma) [inline elements/howto-menu/demo.html]: Any reason why you are choosing `visibility: hidden` over `display: none`?

File: elements/howto-menu/demo.html [hunk 2]
```diff
@@ -0,0 +1,43 @@
+<!doctype html>
+
+<style>
+  /* The tooltip is by default unstyled. */
+  howto-menu {
+    display: block;
```
Comment (surma) [inline elements/howto-menu/demo.html]: With the implicit `position: static` the menu will push down following content when opened and re-layout the entire page. I think `position: absolute` & some JavaScript to set the position might be worth a try.

File: elements/howto-menu/demo.html [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  }
+
+  /* The menu is open. */
+  howto-menu[aria-hidden="false"] {
+    visibility: visible;
+  }
+</style>
+
+<div class="text">
+   <button aria-label="menu" id="menu-btn">open menu</button>
```
Comment (surma) [inline elements/howto-menu/demo.html]: I don’t think a `<button>` needs a label.

File: elements/howto-menu/demo.html [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+   <howto-menu aria-labelledby="menu-btn" aria-hidden="true">
```
Comment (surma) [inline elements/howto-menu/demo.html]: For progressive enhancement it is _probably_ better to have the element set the `aria-hidden` on boot rather than putting it in the markup. When JavaScript is not enabled, the menu will be hidden from screenreaders.

File: elements/howto-menu/demo.html [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+     <howto-menuitem>second option</howto-menuitem>
+     <howto-menuitem>other option</howto-menuitem>
+     <howto-menuitem>yet another option</howto-menuitem>
+   </howto-menu>
+</div>
+
+<script>
+  document.getElementById('menu-btn').addEventListener('click', e => {
+    var menu = document.querySelector('howto-menu');
+    menu.toggle();
```
Comment (surma) [inline elements/howto-menu/demo.html]: So far we have been implementing over APIs with attributes as it allows to integrate with frameworks like Angular well. So instead you’d do
```js
if (menu.hasAttribute('expanded')
  menu.removeAttribute('expanded');
else
  menu.setAttribute('expanded', '');
```

or something along those lines.

File: elements/howto-menu/howto-menu.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+ * A menu is a widget that offers a list of choices to the user,
+ * such as a set of actions or functions. A menu is usually opened,
+ * or made visible, by activating a menu button, choosing an item in a menu
+ * that opens a sub menu, or by invoking a command, such as Shift + F10 in
+ * Windows, that opens a context specific menu.
+ *
+ * The element that opens the menu is referenced with aria-labelledby.
+ * TODO(devnook): Should it be aria-labelledby or aria-describedby?
+ *
+ * See: https://www.w3.org/TR/wai-aria-practices-1.1/#menu
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: That’s a pretty good idea. We should always put that link there!

Comment (surma) [inline elements/howto-menu/howto-menu.js]: Opened #48

File: elements/howto-menu/howto-menu.js [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+
+ class HowtoMenu extends HTMLElement {
+
+   /**
+    * The constructor does work that needs to be executed _exactly_ once.
+    */
+   constructor() {
+     super();
+
+     this.KEY_CODES_ = {
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: So far we’ve been putting our code in an IIFE and have the key codes outside the class ([example](https://github.com/GoogleChrome/howto-components/blob/f9f855beeaade5d3d01c0e9f91b861bbf092fe5b/elements/howto-radio-group/howto-radio-group.js#L28))

File: elements/howto-menu/howto-menu.js [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+       ARROW_DOWN: 40,
+       ARROW_UP: 38,
+       ESCAPE: 27,
+     };
+     this.children_ = null;
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Private fields have their underscore at the start.

Comment (surma) [inline elements/howto-menu/howto-menu.js]: We are trying to be stateless, so we have been avoiding variables like this that “cache” children. Currently the menu would not know about dynamically added menu points.

It’s okay to query for all the children whenever you need ([example](https://github.com/GoogleChrome/howto-components/blob/f9f855beeaade5d3d01c0e9f91b861bbf092fe5b/elements/howto-radio-group/howto-radio-group.js#L156)). Alternatively, you could set up a MutationObserver, although I don’t see the need for that complexity.

File: elements/howto-menu/howto-menu.js [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+     // TODO: Consider better naming?
+     this.parent_ = document.getElementById(
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Same.

Regarding naming: “Referencee”? “Opener”?

File: elements/howto-menu/howto-menu.js [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+   }
+
+   moveFocus_(fromEl, toEl) {
+     fromEl.setAttribute('tabindex', -1);
+     toEl.setAttribute('tabindex', 0);
+     toEl.focus();
+   }
+
+   handleBlur_(e) {
+     e.target.setAttribute('tabindex', -1);
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: I am not sure if `blur` bubbles, but if someone does crazy stuff like putting an `<input>` in their menu, this might be do unexpected things. I think a check if `e.target` is actually a `<howto-menuitem>` should suffice.

File: elements/howto-menu/howto-menu.js [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+           }
+         }
+         break;
+       case e.keyCode === this.KEY_CODES_.ESCAPE:
+         // If escape, exit the menu.
+         el.setAttribute('tabindex', -1);
+         this.parent_.focus();
+         this.toggle();
+         break;
+       // TODO: What should Tab key do?
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Should be left to the browser. All keypresses handled by you should call `e.preventDefault()`.

File: elements/howto-menu/howto-menu.js [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+         this.getAttribute('aria-labelledby'));
+   }
+
+   moveFocus_(fromEl, toEl) {
+     fromEl.setAttribute('tabindex', -1);
+     toEl.setAttribute('tabindex', 0);
+     toEl.focus();
+   }
+
+   handleBlur_(e) {
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Call it `event`.

File: elements/howto-menu/howto-menu.js [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+     e.target.setAttribute('tabindex', -1);
+   }
+
+   handleKeyDown_(e) {
+     let el = e.target;
+     switch (true) {
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: I think I prefer a `switch` on `e.keyCode` and handling the letter key range in a separate `if` before the switch.

File: elements/howto-menu/howto-menu.js [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+       case e.keyCode === this.KEY_CODES_.ARROW_DOWN:
+         // If arrow down, move to next item. Wrap if necessary.
+         let next = el.nextElementSibling || this.firstElementChild;
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Move this into a getter á la [this](https://github.com/GoogleChrome/howto-components/blob/f9f855beeaade5d3d01c0e9f91b861bbf092fe5b/elements/howto-radio-group/howto-radio-group.js#L197)

File: elements/howto-menu/howto-menu.js [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+     this.setAttribute('aria-hidden', !makeOpen);
+     if (makeOpen) {
+       // Set focus to the first item.
+       this.firstElementChild.setAttribute('tabindex', 0);
+       this.firstElementChild.focus();
+     }
+   }
+ }
+
+ window.customElements.define('howto-menu', HowtoMenu);
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Where is `howto-menuitem`?

File: elements/howto-menu/demo.html [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+  /* The menu is open. */
+  howto-menu[aria-hidden="false"] {
+    visibility: visible;
+  }
+</style>
+
+<div class="text">
+   <button aria-label="menu" id="menu-btn">open menu</button>
+   <howto-menu aria-labelledby="menu-btn" aria-hidden="true">
+     <howto-menuitem>first option</howto-menuitem>
```
Comment (surma) [inline elements/howto-menu/demo.html]: This CE should be called `<howto-menu-item>`

File: elements/howto-menu/howto-menu.js [hunk 17]
```diff
@@ -0,0 +1,115 @@
...
+     });
+     this.removeEventListener('keydown', this.handleKeyDown_);
+   }
+
+   /**
+    * Open the menu if it was closed and vice versa.
+    */
+   toggle() {
+     // If the menu is hidden, show it.
+     let makeOpen = this.getAttribute('aria-hidden') === 'true';
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Probably nice to have a (private) method with a meaningful name for this.

File: elements/howto-menu/howto-menu.js [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+     this.setAttribute('aria-hidden', !makeOpen);
+     if (makeOpen) {
+       // Set focus to the first item.
+       this.firstElementChild.setAttribute('tabindex', 0);
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: I’d prefer an `_open()` and a `_close()` method that you call into.

File: elements/howto-menu/demo.html (hunk 2 above)
Comment (devnook) [inline elements/howto-menu/demo.html]: I wanted to keep the elements as minimal as possible, therefore was not really styling them (I focused on accessibility first). But you're right, if we want to make the elements performant too they should have some visuals too:) I'd say let's do it in next PR, together with animating etc.

File: elements/howto-menu/demo.html (hunk 4 above)
Comment (devnook) [inline elements/howto-menu/demo.html]: My thinking was that the developer might want to choose whether a menu is visible on the page initially or hidden away. This said, now that I read again the wai-aria guidelines, a visible menu would be rather called a "menubar"... So it would be a different element I guess.

File: elements/howto-menu/howto-menu.js (hunk 9 above)
Comment (devnook) [inline elements/howto-menu/howto-menu.js]: Opener yay!

File: elements/howto-menu/howto-menu.js (hunk 15 above)
Comment (devnook) [inline elements/howto-menu/howto-menu.js]: Removed

**Comment by devnook** [comment]
PTAL. 

Most of the issues are addressed:) Let's add more styling/animations in a next PR. 
Regarding branches, even as a collaborator, I don't think I have write access to create branches in the repo (I get 403).

**Comment by surma** [comment]
That’s because I lied and you didn’t have access >.> I forgot to check, sorry. Fixed now :)

File: elements/howto-menu/howto-menu.js [hunk 19]
```diff
@@ -0,0 +1,216 @@
...
+
+  class HowtoMenu extends HTMLElement {
+
+    /**
+     * The constructor does work that needs to be executed _exactly_ once.
+     */
+    constructor() {
+      super();
+      this._handleBlur = this._handleBlur.bind(this);
+      this._opener = document.getElementById(
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: You only use `_opener` in `_close`, so I think you should just query for it there instead of caching it here.

File: elements/howto-menu/howto-menu.js [hunk 20]
```diff
@@ -0,0 +1,216 @@
...
+         this.getAttribute('aria-labelledby'));
+    }
+
+    static get observedAttributes() {
+      return ['aria-hidden'];
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Rob and I have been trying to use ARA just as an implementation detail. Make your own attributes (`expanded`?) and set `aria-hidden` accordingly instead.

File: elements/howto-menu/howto-menu.js [hunk 21]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+
+    /**
+     * A getter for the first child which is a menuitem.
+     */
+    get firstMenuItem() {
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Should probably be private

File: elements/howto-menu/howto-menu.js [hunk 22]
```diff
@@ -0,0 +1,216 @@
...
+      return this.querySelector('[role^="menuitem"]:first-of-type');
+    }
+
+    /**
+     * A getter for the last child which is a menuitem.
+     */
+    get lastMenuItem() {
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Same

File: elements/howto-menu/howto-menu.js [hunk 23]
```diff
@@ -0,0 +1,216 @@
...
+      return this.querySelector('[role^="menuitem"]:last-of-type');
+    }
+
+    /**
+     * Moves the browser focus from one element to another.
+     * @param {Element} fromEl Element to have focus removed.
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Hm, we haven’t used proper JSDoc so far. Since my parser currently doesn’t understand it, I am not sure it adds value. The names are pretty descriptive, too. @robdodson WDYT?

File: elements/howto-menu/howto-menu.js [hunk 24]
```diff
@@ -0,0 +1,216 @@
...
+
+    /**
+     * A getter for the last child which is a menuitem.
+     */
+    get lastMenuItem() {
+      return this.querySelector('[role^="menuitem"]:last-of-type');
+    }
+
+    /**
+     * Moves the browser focus from one element to another.
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: Probably worth mentioning “roving tabindex” here.

File: elements/howto-menu/howto-menu.js [hunk 25]
```diff
@@ -0,0 +1,216 @@
...
+
+       default:
+         break;
+      };
+      // If letter key, move to an item which starts with that letter.
+      if (event.keyCode > 64 && event.keyCode < 91) {
+        for (let i = 0, child; child = this.children[i]; i++) {
+          if (child.innerText.trim()[0] === event.key) {
+            event.preventDefault();
+            this._moveFocus(el, child);
```
Comment (surma) [inline elements/howto-menu/howto-menu.js]: and `break`, no?

File: elements/howto-menu/demo.html [hunk 26]
```diff
@@ -0,0 +1,39 @@
...
+
+  /* The menu is closed. */
+  howto-menu[aria-hidden="true"] {
+    display: none;
+  }
+</style>
+
+<div class="text">
+   <button id="menu-btn">open menu</button>
+   <howto-menu aria-labelledby="menu-btn">
```
Comment (robdodson) [inline elements/howto-menu/demo.html]: The goal with the label is to help the user understand the purpose of the menu. If you want to continue using a button I'd suggest changing the text to `"options"` instead of `"open menu"`.

File: elements/howto-menu/howto-menu.js (hunk 20 above)
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: I'd suggest creating a boolean property/attribute pair called `.opened` and `[opened]`.

Using `expanded` makes me think it will also set `aria-expanded` but I don't think we need to do that here. Instead we could reserve `expanded`/`aria-expanded` just for submenu items.

File: elements/howto-menu/howto-menu.js [hunk 27]
```diff
@@ -0,0 +1,216 @@
+/*
+ * A menu is a widget that offers a list of choices to the user,
+ * such as a set of actions or functions. A menu is usually opened,
+ * or made visible, by activating a menu button, choosing an item in a menu
+ * that opens a sub menu, or by invoking a command, such as Shift + F10 in
+ * Windows, that opens a context specific menu.
+ *
+ * The element that opens the menu is referenced with aria-labelledby.
+ * TODO(devnook): Should it be aria-labelledby or aria-describedby?
```
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: In the case of a menu button it should be `aria-labelledby`.

Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: You could additionally add an `aria-describedby` field to help explain the purpose of the menu.

example:

label: "Options"
describedby: "Configuration for your personal twitter account"

File: elements/howto-menu/howto-menu.js [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+  };
+
+  class HowtoMenu extends HTMLElement {
+
+    /**
+     * The constructor does work that needs to be executed _exactly_ once.
+     */
+    constructor() {
+      super();
+      this._handleBlur = this._handleBlur.bind(this);
```
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: We use the convention `_onBlur`

File: elements/howto-menu/howto-menu.js (hunk 21 above)
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: yeah just needs an underscore, `_firstMenuItem`

File: elements/howto-menu/howto-menu.js (hunk 24 above)
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: For reference: https://github.com/GoogleChrome/howto-components/blob/master/elements/howto-radio-group/howto-radio-group.js#L8

File: elements/howto-menu/howto-menu.js (hunk 23 above)
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: yeah i think you can remove this

File: elements/howto-menu/howto-menu.js [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+     * Sets up keyboard interactions for menu and its items.
+     */
+    connectedCallback() {
+      // Make children unfocusable by default.
+      const children = Array.from(this.children);
+      children.forEach(el => {
+        el.setAttribute('tabindex', -1);
+        el.addEventListener('blur', this._handleBlur);
+      });
+      this.setAttribute('role', 'menu');
```
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: Check if role is already set.

```
if (!this.hasAttribute('role'))
  this.setAttribute('role', 'menu');
```

File: elements/howto-menu/howto-menu.js [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+        }
+        prev = prev.previousElementSibling;
+      }
+      return null;
+    }
+
+    /**
+     * Opens the menu.
+     */
+    _open() {
```
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: I'd move this, and the work done in `_close` into the `attributeChangedCallback` after you create the `opened` attribute/property.

Take a look at how checkbox does it: https://github.com/GoogleChrome/howto-components/blob/master/elements/howto-checkbox/howto-checkbox.js#L122

That's the pattern we're trying to stick to. Property setters and getters just reflect to attributes and any side effects take place in the `attributeChangedCallback`. This is what Chrome eng suggested we do.

File: elements/howto-menu/howto-menu.js [hunk 31]
```diff
@@ -0,0 +1,216 @@
...
+      }
+    }
+
+    /**
+     * Sets up keyboard interactions for menu and its items.
+     */
+    connectedCallback() {
+      // Make children unfocusable by default.
+      const children = Array.from(this.children);
+      children.forEach(el => {
```
Comment (robdodson) [inline elements/howto-menu/howto-menu.js]: I think having a `reset()` method similar to dash-tabs might be good here.
https://github.com/GoogleChrome/howto-components/blob/master/elements/howto-tabs/howto-tabs.js#L202

That way we could be consistent across elements.

File: elements/howto-menu/howto-menu.e2etest.js [hunk 32]
```diff
@@ -0,0 +1,123 @@
...
+        return document.activeElement === document.querySelector('#menu-btn');
+      `);
+      expect(focusedMenuBtn).to.equal(true);
+      const closedMenu = await this.driver.executeScript(`
+        const menu = document.querySelector('howto-menu');
+        return menu.getAttribute('aria-hidden') === 'true';
+      `);
+      expect(closedMenu).to.equal(true);
+    }
+  );
```
Comment (robdodson) [inline elements/howto-menu/howto-menu.e2etest.js]: Also need to make sure the menu closes when the user tabs out of it.

File: elements/howto-menu/demo.html [hunk 33]
```diff
@@ -0,0 +1,39 @@
...
+     <span role="menuitem">first option</span>
+     <span role="menuitem">second option</span>
+     <span role="menuitem">other option</span>
```
Comment (robdodson) [inline elements/howto-menu/demo.html]: I'd really like to see an example of at least one nested menu item that uses `aria-haspopup`.

In general we've tried to avoid developers needing to write out ARIA attributes themselves, so it might be a good use case to create a `<howto-menu-item>` element that automatically self applies `aria-haspopup` if it has children.

File: elements/howto-menu/howto-menu.js (hunk 23 above)
Comment (devnook) [inline elements/howto-menu/howto-menu.js]: Sorry, old habits from my Closure days;)

File: elements/howto-menu/howto-menu.js (hunk 20 above)
Comment (devnook) [inline elements/howto-menu/howto-menu.js]: How about using built in 'hidden' attribute instead? Do you think it would be too much overload on the term?

File: elements/howto-menu/demo.html (hunk 33 above)
Comment (devnook) [inline elements/howto-menu/demo.html]: I plan to add menuitem, menuitemradio and menuitemcheckbox. Also, there should be a demo where all of them work together. Is it ok to do it in a separate PR though?

Comment (surma) [inline elements/howto-menu/demo.html]: sg

File: elements/howto-menu/howto-menu.js (hunk 30 above)
Comment (devnook) [inline elements/howto-menu/howto-menu.js]: I'm not sure I got it correctly - you want the body of the _open and _close to be moved back  to  attributeChangedCallback? They are already called from there, just with the 'hidden' property and not 'opened'.

**Comment by devnook** [comment]
PTAL

Most issues are addressed, apart from attributeChangedCallback (need a clarification from @robdodson there).

Also, tests seem broken, but it happens also on master - do you know where's the issue?

**Comment by surma** [comment]
Master works for me (and Travis) just fine. Since this is in your repo, I can’t fix it for you, but a rebase to master should do it :)

**Comment by devnook** [comment]
I'm abandoning this PR in favour of #78 (Somehow I'm unable to fix tests on this branch).
